src/components/tooltip/tooltip.svelte

```diff
@@ -0,0 +1,175 @@
+<script lang="ts" context="module">
+  export const modes = ['hero', 'info', 'default', 'mini'] as const
+  export type Mode = (typeof modes)[number]
+</script>
+
+<script lang="ts">
+  import type { MiddlewareData, Placement } from '@floating-ui/dom'
+  import { arrow as arrowMiddleware } from '@floating-ui/dom'
+  import { createEventDispatcher } from 'svelte'
+  import { fade } from 'svelte/transition'
+  import Floating from '../floating/floating.svelte'
+
+  export let text: string = undefined
+
+  export let placement: Placement = 'top'
+  export let flip: boolean = true
+  export let shift: number | undefined = 8
+  export let offset: number = 8
+  export let mode: Mode = 'mini'
+
+  /* Whether the tooltip is currently visible */
+  export let visible: boolean | undefined = undefined
+
+  // Note: This is separate from the |visible| flag because we want to handle
+  // controlled and uncontrolled states for this component.
+  $: visibleInternal = visible ?? false
+
+  const dispatch = createEventDispatcher()
+
+  let tooltip: HTMLElement
+  let arrow: HTMLElement
+  let trigger: HTMLElement
+
+  function positionArrow(
+    e: CustomEvent<{ middlewareData: MiddlewareData; placement: Placement }>
+  ) {
+    const { x: arrowX, y: arrowY } = e.detail.middlewareData.arrow
+
+    const staticSide = {
+      top: 'bottom',
+      right: 'left',
+      bottom: 'top',
+      left: 'right'
+    }[e.detail.placement.split('-')[0]]
+
+    if (!arrow) return
+
+    Object.assign(arrow.style, {
+      left: arrowX != null ? `${arrowX}px` : '',
+      top: arrowY != null ? `${arrowY}px` : '',
+      right: '',
+      bottom: '',
+      [staticSide]: '-4px'
+    })
+  }
+
+  function setVisible(newVisible: boolean) {
+    if (newVisible === visible) return
+
+    if (visible === undefined) visibleInternal = newVisible
+    dispatch('visibilitychange', { visible: newVisible })
+  }
+</script>
+
+<div
+  class="leo-tooltip"
+  on:mouseenter={() => setVisible(true)}
+  on:mouseleave={() => setVisible(false)}
+  on:focusin={() => setVisible(true)}
+  on:focusout={() => setVisible(false)}
+>
+  {#key visibleInternal}
+    <Floating
+      target={trigger}
+      {flip}
+      {offset}
+      {placement}
+      {shift}
+      autoUpdate
+      middleware={[arrowMiddleware({ padding: 0, element: arrow })]}
+      on:computedposition={positionArrow}
+    >
+      <div
+        class="tooltip"
+        class:hero={mode === 'hero'}
+        class:info={mode === 'info'}
+        class:mini={mode === 'mini'}
+        class:default={mode === 'default' || !mode}
+        transition:fade={{ duration: 60 }}
+        hidden={!visibleInternal}
```

Comment:
Can we add aria attributes for hidden please

Reply:
Actually maybe we don't need that because the `hidden` attribute seems to be making the content hidden from the accessibility tree